### ISSUE TITLE
At mgm.py:67, the gradients computed (in the 'output' variable) are all zero.

### ISSUE THREAD

**Issue by nathanbollig** [nathanbollig/mgm#2]
**Technical investigation**

Model structure is:
Conv1d -> (2392, 32)
MaxPool1D -> (478, 32)
Flattened to 15296: values on the order of 0.01, + and -
Dense 16: all zeros
Dense 1: 0.5844933

The outputs from the second to last dense layer (16 values) are all zero for this input. Therefore the sigmoid activation is applied to the bias term of the final output layer.

bias of final dense layer = 0.34124666
applying sigmoid activation = 0.5844933

The outputs of the flattened layer are not all zero. This is not a vanishing gradient issue in the sense of final activation being very large.

Let's investigate the first hidden unit calculation of the dense 16 layer:

```
# Get weights for the first hidden unit
dense16_weights = model.layers[-2].weights
b = dense16_weights[1][0]
w = dense16_weights[0][:,0]

# Decapitate model so that flattened layer is its final output
decapitated_model = model.layers[-3].output
model_flattened_layer = tf.keras.Model(inputs=model.inputs, outputs=decapitated_model)

# Get outputs the flattened layer for this input
input_dtype = model.dtype
input = tf.Variable(seq.to_predict(representation), dtype=getattr(tf, input_dtype))
flatten_outputs = model_flattened_layer(input)

# Do hidden unit calculation for the first unit of the next layer
flatten_outputs = tf.reshape(flatten_outputs, (1,-1))
w = tf.reshape(w, (-1,1))
hidden1 = tf.matmul(flatten_outputs, w) + b
```

hidden1 is -0.90106463, which passes through ReLU and becomes 0

```
# Compute pre-activation values for other hidden units
for i in range(16):
	b = dense16_weights[1][i]
	w = tf.reshape(dense16_weights[0][:,i], (-1,1))
	print(tf.matmul(flatten_outputs, w) + b)
```
	
Gives the following:
```
tf.Tensor([[-0.90106463]], shape=(1, 1), dtype=float32)
tf.Tensor([[-10.381929]], shape=(1, 1), dtype=float32)
tf.Tensor([[-11.434735]], shape=(1, 1), dtype=float32)
tf.Tensor([[-0.40763256]], shape=(1, 1), dtype=float32)
tf.Tensor([[-9.928726]], shape=(1, 1), dtype=float32)
tf.Tensor([[-0.63950086]], shape=(1, 1), dtype=float32)
tf.Tensor([[-9.900103]], shape=(1, 1), dtype=float32)
tf.Tensor([[-0.48936638]], shape=(1, 1), dtype=float32)
tf.Tensor([[-8.898059]], shape=(1, 1), dtype=float32)
tf.Tensor([[-12.527808]], shape=(1, 1), dtype=float32)
tf.Tensor([[-13.750477]], shape=(1, 1), dtype=float32)
tf.Tensor([[-5.516201]], shape=(1, 1), dtype=float32)
tf.Tensor([[-9.332866]], shape=(1, 1), dtype=float32)
tf.Tensor([[-12.4929285]], shape=(1, 1), dtype=float32)
tf.Tensor([[-6.515724]], shape=(1, 1), dtype=float32)
tf.Tensor([[-5.80045]], shape=(1, 1), dtype=float32)
```

So apparently by luck all the activations are negative, meaning the outputs for all hidden units after ReLU are 0.

Test if gradient calculation changes if this behavior is modified
```

# Get weights
weights = model.layers[-2].get_weights()

# modify a bias term to force a positive activation value
weights[1][0] = 10

# Set weights
model.layers[-2].set_weights(weights)

# Test changes
model(input, training=False)

with tf.GradientTape() as tape:
	target_value = model(input, training=False)  # Logits for this minibatch
	grads = tape.gradient(target_value, input)

# See output of dense 16
decapitated_model = model.layers[-2].output
model_dense16_layer = tf.keras.Model(inputs=model.inputs, outputs=decapitated_model)
input_dtype = model.dtype
input = tf.Variable(seq.to_predict(representation), dtype=getattr(tf, input_dtype))
dense16_outputs = model_dense16_layer(input)
```

Now gradients sum to 0.2731
Dense 16 outputs are
```
array([[9.096565, 0.      , 0.      , 0.      , 0.      , 0.      ,
        0.      , 0.      , 0.      , 0.      , 0.      , 0.      ,
        0.      , 0.      , 0.      , 0.      ]], dtype=float32)
```

The bias term caused the activations going into the first hidden unit to no longer be negative, and are not zeroed out by the ReLU. This appears to allow for the gradient computation to give a nonzero result.

**Proposed result:** Use leaky ReLU or ELU instead to prevent this chance event of all zero activations.

**Comment by nathanbollig** [comment]
Include handling of zero gradients / zero objective values, since we can't always control the model.
In the deterministic case, current behavior is to select the first zero (non-infinite) index as the minimum objective value.

New behavior:
- Throw warning if gradients are all zero
- Throw warning if all non-infinite objective values are zero
- Take a random step when all objective values are zero, store comment in `substitution_data`

If the problem occurs frequently, then we can investigate ELU or tanh activation functions to prevent the likelihood of all zero gradients.

**Comment by nathanbollig** [comment]
No warnings upon first test. Investigating when random selection occurred...

```
import pickle
import pandas as pd
from mgm.common.utils import set_data_directory
set_data_directory("spillsim_9-29-22")

with open('variants_list.pkl', 'rb') as fp:
    variants_list = pickle.load(fp)

mgm_d_vars = []
mgm_sas_vars = []
for tup in variants_list:
    mgm_d_var, mgm_sas_var, _, _ = tup
    mgm_d_vars.append(mgm_d_var)
    mgm_sas_vars.append(mgm_sas_var)

for var in mgm_d_vars:
    for sub_dict in var.substitution_data:
        if 'selection_comment' in sub_dict.keys():
            print(var)

for vars in mgm_sas_vars :
    for var in vars:
        for sub_dict in var.substitution_data:
            if 'selection_comment' in sub_dict.keys():
                print(var)
```

No results printed.
